### PR TITLE
Added Weave class and WeaveModel class

### PR DESCRIPTION
**Pull request by NimishaDey**
## Description

Added Weave class and WeaveModel class.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by rbharath — COMMENTED** [review]
@ARY2260 Can you do a preliminary round of review?

File: deepchem/models/torch_models/weavemodel_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+        Iterator which walks over the batches
+        """
+        for epoch in range(epochs):
+            for (X_b, y_b, w_b,
+                 ids_b) in dataset.iterbatches(batch_size=self.batch_size,
+                                               deterministic=deterministic,
+                                               pad_batches=pad_batches):
+                if y_b is not None:
+                    if self.model.mode == 'classification':
+                        y_b = to_one_hot(y_b.flatten(),
```
Comment (ARY2260) [inline deepchem/models/torch_models/weavemodel_pytorch.py:588]: separate handling of labels in the data with in the default generator may not be required

Comment (ARY2260) [reply deepchem/models/torch_models/weavemodel_pytorch.py:588]: please check

File: deepchem/models/torch_models/weavemodel_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+                self.layer.weight_stddev = weight_stddev
+                self.layer.bias_const = bias_const
+                self.layer.dropout = nn.Dropout(dropout)
+                self.layer.layer_act = activation_fn
+                self.layers2.append(self.layer)
+
+        n_tasks = self.n_tasks
+        if self.mode == 'classification':
+            n_classes = self.n_classes
+            self.layer_2 = nn.LazyLinear(n_tasks * n_classes)
```
Comment (ARY2260) [inline deepchem/models/torch_models/weavemodel_pytorch.py:248]: @rbharath will it be fine to use lazy linear over default nn.linear here. The usage here is based on the fact that input size is not known, but it may be possible to get that.

File: deepchem/models/torch_models/weavemodel_pytorch.py (hunk 1 above)
Comment (NimishaDey) [reply deepchem/models/torch_models/weavemodel_pytorch.py:588]: default_generator function gets called internally for Numpy Dataset when model.fit() is run. And to keep it close to tensorflow implementation I thought it would be best to keep the default generator function.

File: deepchem/models/torch_models/weavemodel_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+
+        if mode not in ['classification', 'regression']:
+            raise ValueError(
+                "mode must be either 'classification' or 'regression'")
+
+        regularization_loss: Optional[Callable]
+        if weight_decay_penalty != 0.0:
+            weights = [layer.weight for layer in self.model.layers2]
+            if weight_decay_penalty_type == 'l1':
+                regularization_loss = lambda: weight_decay_penalty * torch.sum(  # noqa: E731
```
[the remark this answers is not in the document]
Comment (NimishaDey) [reply deepchem/models/torch_models/weavemodel_pytorch.py:466]: #noqa has to be used for lambda and since regularization_loss is a callable type I think it's required.

File: deepchem/models/torch_models/weavemodel_pytorch.py [hunk 4]
```diff
@@ -172,7 +171,8 @@ def __init__(
         ]
         self.batch_normalize: bool = batch_normalize
         self.n_weave: int = n_weave
-        torch.manual_seed(21)
+
+        # torch.manual_seed(21)
```
Comment (ARY2260) [inline deepchem/models/torch_models/weavemodel_pytorch.py:175]: is seed not required here?

File: deepchem/models/tests/test_weavemodel_pytorch.py [hunk 5]
```diff
@@ -340,3 +295,77 @@ def test_weave_model_reload():
     original_predict = original_model.predict(dataset)
     reloaded_predict = reloaded_model.predict(dataset)
     assert np.all(original_predict == reloaded_predict)
+
+
+@pytest.mark.torch
+def test_weave_singletask_classification_overfit():
+    """Test weave model overfits tiny data."""
+    # np.random.seed(123)
+    # torch.manual_seed(123)
```
Comment (ARY2260) [inline deepchem/models/tests/test_weavemodel_pytorch.py:304]: I think seed should be turned on for this test.

File: deepchem/models/tests/test_weavemodel_pytorch.py [hunk 6]
```diff
@@ -340,3 +295,77 @@ def test_weave_model_reload():
...
+
+    batch_size = 10
+    model = WeaveModel(n_tasks,
+                       batch_size=batch_size,
+                       learning_rate=0.0003,
+                       dropout=0.0,
+                       mode="classification")
+
+    # Fit trained model
+    model.fit(dataset, nb_epoch=300)
```
Comment (ARY2260) [inline deepchem/models/tests/test_weavemodel_pytorch.py:327]: How long does it take for the test to run for 300 epochs?

File: deepchem/models/torch_models/weavemodel_pytorch.py [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+        n_tasks = self.n_tasks
+        if self.mode == 'classification':
+            n_classes = self.n_classes
+            self.layer_2 = nn.Linear(fully_connected_layer_sizes[1],
+                                     n_tasks * n_classes)
+
+        else:
+            self.layer_2 = nn.Linear(fully_connected_layer_sizes[1], n_tasks)
+
+    def forward(self, inputs: OneOrMany[torch.Tensor]) -> List[torch.Tensor]:
```
Comment (ARY2260) [inline deepchem/models/torch_models/weavemodel_pytorch.py:260]: please add docstrings

File: deepchem/models/torch_models/weavemodel_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+        n_tasks = self.n_tasks
+        if self.mode == 'classification':
+            n_classes = self.n_classes
+            logits: torch.Tensor = torch.reshape(self.layer_2(output),
+                                                 (-1, n_tasks, n_classes))
+            output = F.softmax(logits, dim=2)
+            outputs = [output, logits]
+        else:
+            output = self.layer_2(output)
+            outputs = [output]
```
Comment (ARY2260) [inline deepchem/models/torch_models/weavemodel_pytorch.py:308]: add type annotations here

File: deepchem/models/tests/test_weavemodel_pytorch.py [hunk 9]
```diff
@@ -340,3 +295,77 @@ def test_weave_model_reload():
...
+
+    # Eval model on train
+    scores = model.evaluate(dataset, [classification_metric])
+
```
[the remark this answers is not in the document]
Comment (NimishaDey) [reply deepchem/models/tests/test_weavemodel_pytorch.py:331]: Okay I will do that. The unit test for tensorflow code uses the same value though.

File: deepchem/models/torch_models/weavemodel_pytorch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NimishaDey) [reply deepchem/models/torch_models/weavemodel_pytorch.py:175]: I have readded the seed statement because otherwise the reload test is failing.

[the remark this answers is not in the document]
Comment (NimishaDey) [reply deepchem/models/torch_models/weavemodel_pytorch.py:175]: Is it fine to add the seed statement here?

File: deepchem/models/tests/test_weavemodel_pytorch.py (hunk 5 above)
Comment (NimishaDey) [reply deepchem/models/tests/test_weavemodel_pytorch.py:304]: Done

File: deepchem/models/tests/test_weavemodel_pytorch.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (NimishaDey) [reply deepchem/models/tests/test_weavemodel_pytorch.py:331]: Done

File: deepchem/models/tests/test_weavemodel_pytorch.py (hunk 6 above)
Comment (NimishaDey) [reply deepchem/models/tests/test_weavemodel_pytorch.py:327]: Around 2 mins

File: deepchem/models/torch_models/weavemodel_pytorch.py (hunk 7 above)
Comment (NimishaDey) [reply deepchem/models/torch_models/weavemodel_pytorch.py:260]: Done

File: deepchem/models/torch_models/weavemodel_pytorch.py (hunk 8 above)
Comment (NimishaDey) [reply deepchem/models/torch_models/weavemodel_pytorch.py:308]: Done

File: deepchem/models/torch_models/weavemodel_pytorch.py [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+
+        else:
+            self.layer_2 = nn.Linear(fully_connected_layer_sizes[1], n_tasks)
+
+    def forward(self, inputs: OneOrMany[torch.Tensor]) -> List[torch.Tensor]:
+        """
+        Parameters
+        ----------
+        inputs: OneOrMany[torch.Tensor]
+        Should contain 5 tensors [atom_features, pair_features, pair_split, atom_split, atom_to_pair]
```
Comment (rbharath) [inline deepchem/models/torch_models/weavemodel_pytorch.py:265]: Formatting here is a little off

**Review by rbharath — APPROVED** [review]
LGTM
